File: tobas_core/tobas_rqt_tools/src/tobas_rqt_tools/layouts/scroll_area.py
from overrides import override
from PyQt5.QtWidgets import QLayout, QWidget, QVBoxLayout

from ..widgets import ScrollArea


class ScrollableVBoxLayout(QVBoxLayout):
    def __init__(self) -> None:
        super().__init__()

        scroll_area = ScrollArea()
        super().addWidget(scroll_area)

        self._rows = QVBoxLayout()
        scroll_area.setLayout(self._rows)

    @override
    def addWidget(self, widget: QWidget) -> None:
        self._rows.addWidget(widget)

    @override
    def addLayout(self, layout: QLayout) -> None:
        self._rows.addLayout(layout)

    @override
    def addStretch(self) -> None:
        self._rows.addStretch()

File: tobas_gui/tobas_qt_tools/src/layouts/scroll_area.cpp
#include "tobas_qt_tools/layouts/scroll_area.hpp"

namespace qt
{
ScrollableVBoxLayout::ScrollableVBoxLayout(QWidget* parent) : QVBoxLayout(parent)
{
  auto scroll_area = new ScrollArea();
  addWidget(scroll_area);
  rows_ = new QVBoxLayout(scroll_area);
}

void ScrollableVBoxLayout::addWidget(QWidget* widget)
{
  rows_->addWidget(widget);
}

void ScrollableVBoxLayout::addLayout(QLayout* layout)
{
  rows_->addLayout(layout);
}

void ScrollableVBoxLayout::addStretch()
{
  rows_->addStretch();
}
}  // namespace qt
